File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
def solution(numbers):
    answer = 0
    chae = [True] * 10**7
    chae[0] = False
    chae[1] = False
    for i in range(2, len(chae)):
        if i*i > len(chae):
            break
        if chae[i] == False:
            continue
        for j in range(i*2, len(chae), i):
            chae[j] = False
    #print(chae)
    
    l = len(numbers)
    answer_set = set()
    
    def dfs(old, new, visited_index):
        nonlocal answer
        num = old+new
        num_int = int(num)
        if chae[num_int] and num_int not in answer_set:
            #print(num_int)
            answer += 1
            answer_set.add(num_int)
        
        if len(num) == l:
            return
        
        for i in range(l):
            if i in visited_index:
                continue
            dfs(num, numbers[i], visited_index|{i})
        
        
    for i in range(l):
        if numbers[i] == '0':
            continue
        dfs('', numbers[i], {i})
    
    
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

set<int> candidates; // 소수 후보 저장

bool isPrime(int num){
    if (num < 2) return false;
    for (int i=2;i*i<=num;i++){
        if (num%i == 0) return false;
    }
    return true;
}

void permutation(string current, string remain){
    if (!current.empty()){
        candidates.insert(stoi(current));
        // cout << current << "\n";
    }
    
    // {"", 217} -> {"2", 17}
    for (int i=0;i<remain.size();i++){
        if (current.empty() && remain[i] == '0') continue;
        string next = remain;
        next.erase(i, 1);
        permutation(current + remain[i], next);
    }
}

int solution(string numbers) {
    int answer = 0;
    permutation("", numbers);
    
    for (auto num : candidates){
        if (isPrime(num)) answer++;
    }
    
    return answer;
}